sklearn/manifold/_t_sne.py

```diff
@@ -537,7 +537,7 @@ def trustworthiness(X, X_embedded, *, n_neighbors=5, metric="euclidean"):
     return t
 
 
-class TSNE(BaseEstimator):
+class TSNE(ClassNamePrefixFeaturesOutMixin, TransformerMixin, BaseEstimator):
```

Comment:
I find it weird to have a sub-class of `TransformerMixin` without a `transform` method. But maybe it's better than what we do currently...

Reply:
Since the `Mixin` only define `fit_transform` and that we override it, I find it fine to inherit.

Reply:
The problem we are trying to solve by inheriting is having `TSNE` selected as a transformer by `all_estimators`. Is there another way of doing that, other than having to inherit from `TransformerMixin`?

It feels weird to inherit, which for me is saying "I am a transformer with all the things a transformer can do!" and then not have a `transform` method. But the mixin doesn't define it either. So is having a `transform` method not part of being "a real transformer"?

On a pragmatic level, using the mixin is a nice way to get into the list of `all_estimators` and maybe mixins aren't like real "is a" style inheritance?

Reply:
> "I am a transformer with all the things a transformer can do!"

Looking at the documentation, it seems that the original spirit is indeed to have at least `fit` and `transform` and `fit_transform` is just a convenience.

> On a pragmatic level, using the mixin is a nice way to get into the list of all_estimators and maybe mixins aren't like real "is a" style inheritance?

We could always replace by duck-typing: if an estimator implements `fit`+`transform` or/and `fit_transform`, then it should pass the test of a transformer.
Indeed, some of our checks are ducktyping while the helper to list the estimators is checking for the mixins.

Reply:
Actually, the annoying part with duck-typing is that we already need to have an instance while we are playing with classes at this stage.

Reply:
I don't quite understand what you mean. To find out if a method is implemented you could do something like `"fit" in dir(Estimator)` no?

Reply:
Indeed we could. I just had in my mind a piece of the parameter validation framework `HasMethods` that does this job already for instances.

Reply:
According to the glossary, a transformer is an estimator that implement transform and/or fit_transform, so TSNE complies with that and since ``TransformerMixin`` is supposed to be the "mixing class for all transformers in scikit-learn", then I find it totally appropriate that TSNE inherits from this mixin

Reply:
Good point.

Yet, I am still not convinced we should rely on `isinstance(obj, TransformerMixin)` to discover all transformers in scikit-learn. We should rather use an estimator tag or duck-typing.

But +1 with moving forward with this TSNE-specific PR and delegate the discussion of how to properly discover transformers in common test to another issue/PR.

Reply:
> We should rather use an estimator tag or duck-typing.

This is related to https://github.com/scikit-learn/scikit-learn/pull/17806/files which is about all other estimator types.